### PR TITLE
ocsp: make ocsp request with POST method

### PR DESCRIPTION
**Pull request by wallyqs**
This makes the server send a POST request to the OCSP Server, then falling back to use GET as in previous versions.

### CODE REVIEW

**Comment by derekcollison** [comment]
Why do we need this functionality?

**Comment by philpennock** [comment]
Derek: interoperability with various network setups, such that a GET request might not succeed and a POST gets a cleaner (HTTP-caching-wise) path to the origin and is less tampered with.  Also, RFC 6960 has size limits on the request (since the DER gets base64 encoded into the URL, which can have size limits) so using POST avoids those size limits and can work if using larger RSA keys.

**Comment by derekcollison** [comment]
Should we just support POST vs this weird duality or is that just table stakes for OCSP ecosystem etc?

**Comment by wallyqs** [comment]
POST is more common from what I've seen so we could change to that instead.

File: server/ocsp.go
Comment (philpennock) [inline server/ocsp.go:197]: Shouldn't this be `if resp.StatusCode > 299` to return an error and then perhaps if it's not 200 then log it as being of interest but not an error?

File: server/ocsp.go
Comment (philpennock) [inline server/ocsp.go:233]: If the POST is more reliable and the GET fails too, this loses our POST error response.  Any chance we can use `errors.Join()` to return both errors, instead of risking masking a solid POST error response with a poor HTTP proxy GET error response which doesn't tell us why the POST failed in the first place?

**Review by philpennock — APPROVED** [review]

File: server/ocsp.go
Comment (wallyqs) [reply server/ocsp.go:197]: I will make this change.

File: server/ocsp.go
Comment (wallyqs) [reply server/ocsp.go]: good idea, I applied these fixes.

**Comment by wallyqs** [comment]
Applied the changes so now ready.

File: server/ocsp.go
Comment (derekcollison) [inline server/ocsp.go:166]: Do we need Get as well here? Do all OCSP support POST? Should we just support that?

Comment (philpennock) [reply server/ocsp.go:166]: In theory they should, but the problem is "do they really, and do the applicaiton firewalls permit it", inside corporate environments.  So since we have been using GET, I think the goal is to make sure that worst case scenario, for things we don't control, we don't cause a regression in system behavior.

Comment (wallyqs) [reply server/ocsp.go:166]: They should support both, but we have used GETs only so far so in case there was a firewall or something depending on that behavior already we do the fallback just in case.

File: server/ocsp.go
Comment (philpennock) [inline server/ocsp.go:238]: I think you need an `err = nil` before the break here: if the POST fails but the GET succeeds, we want to proceed.

**Review by philpennock — CHANGES_REQUESTED** [review]

**Review by derekcollison — APPROVED** [review]
LGTM

**Comment by derekcollison** [comment]
@philpennock you had a block on this so re-requested your review to make sure its cleared.

**Comment by philpennock** [comment]
> @philpennock you had a block on this so re-requested your review to make sure its cleared.

Yes, the block is 6 minutes old because it's for a bug introduced in the latest push. :D

**Comment by derekcollison** [comment]
@wallyqs ping me when ready and cleared.

File: server/ocsp.go
Comment (wallyqs) [reply server/ocsp.go:240]: nice catch, it was also failing when trying multiple ocsp servers urls after one failed and the following one succeeded. This is fixed now.

**Comment by derekcollison** [comment]
ok @philpennock let me know when you have LGTM'd.

**Review by philpennock — APPROVED** [review]

**Comment by derekcollison** [comment]
Will wait on green then merge..

**Pull request merged**
